pages/handbook/product-engineering/playbooks/code-structure.mdx

```diff
@@ -0,0 +1,178 @@
+---
+description: Learn how Langfuse's backend is structured and how to write code following our patterns and best practices.
+---
+
+# Backend Code Structure Guide
+
+This guide explains how Langfuse's backend is organized and how to write code that follows our established patterns. 
+
+## Architecture at a Glance
+
+Langfuse uses a **monorepo** structure with three main packages:
+
+- **web** - Next.js 14 application (UI + tRPC API + Public REST API)
+- **worker** - Express-based background job processor using BullMQ
+- **packages/shared** - Shared code, types, and utilities used by both web and worker
+
+### API Request Flow
+
+```
+┌─ Web (NextJs): tRPC API ────┐   ┌── Web (NextJs): Public API ─┐
+│                             │   │                             │
+│  HTTP Request               │   │  HTTP Request               │
+│      ↓                      │   │      ↓                      │
+│  tRPC Procedure             │   │  withMiddlewares +          │
+│  (protectedProjectProcedure)│   │  createAuthedProjectAPIRoute│
+│      ↓                      │   │      ↓                      │
+│  Service (business logic)   │   │  Service (business logic)   │
+│      ↓                      │   │      ↓                      │
+│  Prisma / ClickHouse        │   │  Prisma / ClickHouse        │
+│                             │   │                             │
+└─────────────────────────────┘   └─────────────────────────────┘
+                 ↓
+            [optional]: Publish to Redis BullMQ queue
+                 ↓
+┌─ Worker (Express): BullMQ Queue Job ────────────────────────┐
+│                                                             │
+│  BullMQ Queue Job                                           │
+│      ↓                                                      │
+│  Queue Processor (handles job)                              │
+│      ↓                                                      │
+│  Service (business logic)                                   │
+│      ↓                                                      │
+│  Prisma / ClickHouse                                        │
+│                                                             │
+└─────────────────────────────────────────────────────────────┘
+```
+We follow the layered architecture pattern:
+- Router Layer: HTTP Requests or BullMQ Job handlers
+- Service Layer: Contains all the business logic
```

Comment:
Do you have an example of a well-structured API/Public API here? Please link. 

My perspective on the codebase is that in many places we do not follow this architecture strictly. There are quite some examples (I followed/adapted) that just keep all the logic in the trpc router (historically). 

I think as a new joiner, having an example as "gold standard" would help a lot to get startet.